src/main/java/de/tum/cit/aet/artemis/plagiarism/service/ProgrammingPlagiarismDetectionService.java

```diff
@@ -384,27 +381,133 @@ private List<Repository> downloadRepositories(ProgrammingExercise programmingExe
                 else {
-                    deleteTempLocalRepository(repo);
+                    // If no minimum size is specified, include all repositories
+                    return repo;
                 }
             }
             catch (GitException | GitAPIException | InvalidPathException ex) {
                 log.error("Clone student repository {} in exercise '{}' did not work as expected: {}", participation.getVcsRepositoryUri(), programmingExercise.getTitle(),
                         ex.getMessage());
+                return null;
             }
-        });
+        }).filter(Objects::nonNull).toList();
+
+        downloadedRepositories.addAll(studentRepositories);
+
+        // Update progress message
+        var progressMessage = "Processing repositories: " + downloadedRepositories.size() + " valid out of " + maxRepositories + " total";
+        plagiarismWebsocketService.notifyInstructorAboutPlagiarismState(topic, PlagiarismCheckState.RUNNING, List.of(progressMessage));
+
+        log.info("Downloaded and filtered {} repositories out of {} participations for exercise {} (minimum token size: {} tokens)", downloadedRepositories.size(),
+                participations.size(), programmingExercise.getId(), minimumTokenSize);
 
         return downloadedRepositories;
     }
+
+    /**
+     * Checks if a repository meets the minimum size requirement by counting tokens in relevant files.
+     * Returns true as soon as the minimum token count is reached across all files.
+     * Returns true in case of any errors to be INCLUSIVE. I/O errors should not prevent plagiarism check.
+     *
+     * @param repository          The repository to check
+     * @param programmingExercise The programming exercise
+     * @param minimumTokenSize    The minimum number of tokens required
+     * @return true if the repository meets the minimum size requirement or if there are any errors
+     */
+    private boolean meetsMinimumSize(Repository repository, ProgrammingExercise programmingExercise, int minimumTokenSize) {
+        try {
+            Path repoPath = repository.getLocalPath();
+            if (!Files.exists(repoPath) || !Files.isDirectory(repoPath)) {
+                log.warn("Repository path does not exist or is not a directory: {}", repoPath);
+                return true;
+            }
+
+            // Get file extensions for the programming language for filtering
+            Set<String> fileExtensions = programmingExercise.getProgrammingLanguage().getFileExtensions().stream().map(ext -> "." + ext).collect(Collectors.toSet());
+
+            try (Stream<Path> paths = Files.walk(repoPath)) {
+                List<Path> relevantFiles = paths.filter(Files::isRegularFile).filter(path -> {
+                    // Only consider files with the correct file extension
+                    String fileName = path.getFileName().toString().toLowerCase();
+                    return fileExtensions.stream().anyMatch(fileName::endsWith);
```

Comment:
nit: you can reuse public boolean `matchesFileExtension(String path`) of ProgrammingLanguage so no ned for local var fileExtensions and the own logic for match here 
```suggestion
                   return programmingExercise.getProgrammingLanguage().matchesFileExtension(fileName)
```